MudDataGrid - Multi Level Grouping

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Introduce Multi Level Grouping to MudDataGrid without any breaking changes.
Resolves #7587 
Resolves #5654 
Created two-way bind for GroupExpanded and GroupByOrder
Also consolidated logic for showing items in DataGridVirtualizeRow

![image](https://github.com/user-attachments/assets/7a7ac426-760c-4203-84dc-fc0ffc589ff0)

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Visually, Unit Tests

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

ScarletKuro commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -2160,5 +2249,21 @@ protected virtual void Dispose(bool disposing)
+
+        private sealed class EmptyGrouping<TKey, TElement> : IGrouping<TKey, TElement>
+        {
+            private readonly List<TElement> _elements = new();
```
We should not allocate an empty list in memory if the list is never used.
We should instead dp:
```C#
public IEnumerator<TElement> GetEnumerator() => Enumerable.Empty<TElement>().GetEnumerator();

System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
```
which is free memory wise

ScarletKuro commented on src/MudBlazor/Components/DataGrid/DataGridGroupRow.razor.cs:
```diff
@@ -0,0 +1,94 @@
+        [Category(CategoryTypes.DataGrid.Appearance)]
+        public string? StyleClass { get; set; }
+
+        protected override void OnInitialized()
```
I think that can be removed since there is no custom logic

versile2 replied:
It exists as a parameter on MudDataGrid and I am passing it to DataGridGroupRow and using it in GroupStylename. While I agree that we should get rid of Style won't this be a breaking change?

ScarletKuro commented on src/MudBlazor/Components/DataGrid/DataGridGroupRow.razor.cs:
```diff
@@ -0,0 +1,94 @@
+
+        private void GroupExpandClick()
+        {
+            Expanded = !Expanded;
```
Parameter state please, or private field, we don't use that in new code.

versile2 replied:
Should I use a private _expanded property or ParameterState? Obviously this isn't really an external component so I am not sure if ParameterState is the right call.

versile2 replied:
I went ahead and used a private _expanded property. Let me know if you want ParameterState instead.

ScarletKuro commented on src/MudBlazor/Components/DataGrid/DataGridVirtualizeRow.razor.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        [Parameter]
+        [Category(CategoryTypes.DataGrid.Data)]
+        public required MudDataGrid<T> DataGrid { get; set; }
```
I don't think using `required` is a good idea for Razor components.  
Yes, it works while the framework uses reflection to set those values and bypass this. However, if they switch to using source generators in the future, it will throw an compile time error, since the parameters values won't be set when constructing the object.  
Instead, use the `[EditorRequired]` attribute, which can be recognized by the Razor editor.

ScarletKuro commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor:
```diff
@@ -94,18 +93,18 @@
-                                                        AnchorOrigin="Origin.TopLeft"
-                                                        TransformOrigin="Origin.TopLeft" Elevation="4"
-                                                        Style="max-width:700px;z-index:11;" MaxHeight="400">
+                                            AnchorOrigin="Origin.TopLeft"
```
It seems the intent is wrong now, and not just in that place. I'm actually against any changes to the intent in the datagrid, even if it wasn't initially correct, like in the range of 44-60. This component is too large, and understanding the changes is difficult. I'm struggling to review them.

versile2 replied:
I did this as a single commit so I could remove it if need be, as soon as I'm done with my next task I'll remove commit 
da3f71d8d59fc9822a0fac91363c70e3d98dbfde

ScarletKuro commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor:
```diff
@@ -444,246 +314,243 @@
     internal RenderFragment ToolbarMenu(MudDataGrid<T> dataGrid) =>
-        @<text>
-            <MudMenu Icon="@Icons.Material.Outlined.Settings" Size="Size.Small" Dense="true">
+    @<text>
```
This indentation is not correct, and I'm not sure why your IDE does this.  
The `@<text>` shouldn't be on the same level as `internal`, and if I run the reformat, this section will return to its original state. Even outside of Razor, the previous indentation is more consistent with our current style preferences. And as I mentioned, even if it wasn't, let's keep the original indentation as much as possible for easier review.

ScarletKuro commented:
> I'm happy to see it back on MudDataGrid

We had this before? In v6?

ScarletKuro commented:
Any chance you could review this @tjscience?

henon commented:
@versile2 can this be fixed?

![image](https://github.com/user-attachments/assets/21040626-387a-4eeb-8647-3758577db482)

versile2 commented:
> @versile2 can this be fixed?
> 
> ![image](https://private-user-images.githubusercontent.com/44090/432326105-21040626-387a-4eeb-8647-3758577db482.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQyOTQxMzAsIm5iZiI6MTc0NDI5MzgzMCwicGF0aCI6Ii80NDA5MC80MzIzMjYxMDUtMjEwNDA2MjYtMzg3YS00ZWViLTg2NDctMzc1ODU3N2RiNDgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEwVDE0MDM1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0Y2FhNmQ4MjdlNzI5NmE5YzU3ODllNzU1NzkxOWNmYTlmMWM0Y2Q2MmUxMzg2NzQ3YTAzYWU3YTkxOWI0ZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xgaWgm8csb23rn0zA87awDFCbsTejHLjUgHpkmbuzv0)

Of course, how do you want it?

henon commented:
Perfectly aligned on the base line, if you please, with a bit of padding between Information and 2025 ;)

versile2 commented:
> Perfectly aligned on the base line, if you please, with a bit of padding between Information and 2025 ;)

![image](https://github.com/user-attachments/assets/55e9da08-0cdc-4444-ad1e-9cd139f00dc1)

This ok?

ScarletKuro commented on src/MudBlazor/Components/DataGrid/DataGridGroupRow.razor.cs:
```diff
@@ -0,0 +1,105 @@
+
+        [Parameter]
+        [Category(CategoryTypes.DataGrid.Behavior)]
+        public bool Expanded
```
You did not address the problem with writting to the parameter directly, you can converted from auto property
from auto property:
```
public bool Expanded { get; set; }
```
to the one with backing field:
```
public bool Expanded
{
	get => _expanded;
	set => _expanded = value;
}
```
Which is the same, because it what the compiler turns it into.
In MSFT examples or ParameterState those are independent

versile2 replied:
Removed expanded as a property and commented the Toggle method better

ScarletKuro reviewed (changes requested):
I don’t like that so many existing grouping tests were removed. Shouldn’t this Multi-Level Grouping change be added on top of the existing grouping functionality, so that it’s optional? The old tests that don’t use Multi-Level Grouping should remain in place.

Also, there isn’t a single bUnit test added for this change. The `DataGridGroupingTests` are not used explicitly in bUnit, it just act as visual tests. Unfortunately, I can’t accept such a significant change without any bUnit tests.

versile2 commented:
> I don’t like that so many existing grouping tests were removed. Shouldn’t this Multi-Level Grouping change be added on top of the existing grouping functionality, so that it’s optional? The old tests that don’t use Multi-Level Grouping should remain in place.
> 
> Also, there isn’t a single bUnit test added for this change. The `DataGridGroupingTests` are not used explicitly in bUnit, it just act as visual tests. Unfortunately, I can’t accept such a significant change without any bUnit tests.

1) Every grouping test was moved to a separate file to keep them together. Not one single test was removed, instead I added tests.
2) Every test in that file is a bunit test? I've no idea what you mean.
3) I asked you about the _expanded and you never responded.

ScarletKuro commented:
> * Every grouping test was moved to a separate file to keep them together. Not one single test was removed, instead I added tests.
> 
> * Every test in that file is a bunit test? I've no idea what you mean.
> 
> * I asked you about the _expanded and you never responded.

Oh, so you moved everything to `DataGridGroupingTests`, the file is big and github didn't unfold it, and I just saw that you removed everything from the `DataGridTests`.
I saw the `_expanded` was added, and when you said you used `_expanded` I thought you did the same thing like in MSFT https://learn.microsoft.com/en-us/aspnet/core/blazor/components/overwriting-parameters?view=aspnetcore-9.0 that the parameters are separate and only way they relate to each other is through syncing on `OnParametersSet`.
We just need to address the `Expanded` then and then we can merge it.

versile2 commented:
> > * Every grouping test was moved to a separate file to keep them together. Not one single test was removed, instead I added tests.
> > * Every test in that file is a bunit test? I've no idea what you mean.
> > * I asked you about the _expanded and you never responded.
> 
> Oh, so you moved everything to `DataGridGroupingTests`, the file is big and github didn't unfold it and i scrolled past it, and I just saw that you removed everything from the `DataGridTests`. I saw the `_expanded` was added, and when you said you used `_expanded` I thought you did the same thing like in MSFT https://learn.microsoft.com/en-us/aspnet/core/blazor/components/overwriting-parameters?view=aspnetcore-9.0 that the parameters are separate and only way they relate to each other is through syncing on `OnParametersSet`. We just need to address the `Expanded` then and then we can merge it.

Yes, at 5000 lines the file was becoming real big so I separated. I hope that is okay. 
Adjusted the expanded logic to be what is intended, a trickle down from _groupExpansionsDict.
Modified your changes of Example to include all results when not filtered. 
Copied those changes to the ViewerTest.
Obsoleted ExpandAllGroups and CollapseAllGroups in favor of the Async versions.

versile2 commented:
@ScarletKuro Let me know if you need me to do anything else on this one.

ScarletKuro commented:
> @ScarletKuro Let me know if you need me to do anything else on this one.

There are conflicts that needs to be fixed

ScarletKuro commented:
`GroupExpandClick_ShouldToggleExpandedState` test is flaky, sometimes green and sometimes it gives:
```
public async Task GroupExpandClick_ShouldToggleExpandedState()

Expected dataGrid.Instance._groupExpansionsDict.Count() to be 0, but found 1 (difference of 1).
   at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Numeric.NumericAssertions`2.Be(T expected, String because, Object[] becauseArgs)
   at MudBlazor.UnitTests.Components.DataGridGroupingTests.<>c__DisplayClass13_0.<GroupExpandClick_ShouldToggleExpandedState>g__GetCount|2(Boolean currExpanded) in C:\Users\Kuro\source\repos\MudBlazor\src\MudBlazor.UnitTests\Components\DataGridGroupingTests.cs:line 470
   at MudBlazor.UnitTests.Components.DataGridGroupingTests.GroupExpandClick_ShouldToggleExpandedState() in C:\Users\Kuro\source\repos\MudBlazor\src\MudBlazor.UnitTests\Components\DataGridGroupingTests.cs:line 485
```

versile2 commented:
> `GroupExpandClick_ShouldToggleExpandedState` test is flaky, sometimes green and sometimes it gives:
> 
> ```
> public async Task GroupExpandClick_ShouldToggleExpandedState()
> 
> Expected dataGrid.Instance._groupExpansionsDict.Count() to be 0, but found 1 (difference of 1).
>    at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message)
>    at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
>    at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
>    at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
>    at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
>    at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
>    at FluentAssertions.Numeric.NumericAssertions`2.Be(T expected, String because, Object[] becauseArgs)
>    at MudBlazor.UnitTests.Components.DataGridGroupingTests.<>c__DisplayClass13_0.<GroupExpandClick_ShouldToggleExpandedState>g__GetCount|2(Boolean currExpanded) in C:\Users\Kuro\source\repos\MudBlazor\src\MudBlazor.UnitTests\Components\DataGridGroupingTests.cs:line 470
>    at MudBlazor.UnitTests.Components.DataGridGroupingTests.GroupExpandClick_ShouldToggleExpandedState() in C:\Users\Kuro\source\repos\MudBlazor\src\MudBlazor.UnitTests\Components\DataGridGroupingTests.cs:line 485
> ```

Fixed, I added proper wait for assertions.

![image](https://github.com/user-attachments/assets/708ad21b-7884-4b4c-8341-31ad1d5fa8fc)

ScarletKuro commented:
Still fails for me, now it's
```
Bunit.Extensions.WaitForHelpers.WaitForFailedException : The assertion did not pass within the timeout period.

If this test does not fail consistently, the reason may be that
the wait timeout is too short, and the runtime did not have enough
time to complete the necessary number of renders of the component under test.
This can happen on highly utilized or slower hardware, for example.

To determine if this is the cause, compare the check and render count(s) below
and see if they match what is expected. If they do not,
consider increasing the timeout, either at the individual
method call level, e.g. WaitForElement("div", TimeSpan.FromSeconds(15)),
or via the static TestContext.DefaultWaitTimeout property.

Check count: 1.
Component render count: 3.
Total render count across all components: 8.
  ----> NUnit.Framework.AssertionException : Expected dataGrid.Instance._groupExpansionsDict.Count to be 0, but found 1 (difference of 1).
   at Bunit.RenderedFragmentWaitForHelperExtensions.WaitForAssertion(IRenderedFragmentBase renderedFragment, Action assertion, Nullable`1 timeout) in /_/src/bunit.core/Extensions/WaitForHelpers/RenderedFragmentWaitForHelperExtensions.cs:line 75
   at MudBlazor.UnitTests.Components.DataGridGroupingTests.GroupExpandClick_ShouldToggleExpandedState() in C:\Users\Kuro\source\repos\MudBlazor\src\MudBlazor.UnitTests\Components\DataGridGroupingTests.cs:line 484
```

ScarletKuro commented:
Are you sure it's a good idea to test it by directly invoking `GroupExpandClickAsync`?  
Don’t we have an HTML counterpart with a handler like `<MudIconButton ... OnClick="@GroupExpandClick" />`, so we could just emulate clicking it instead? That’s always better.  
Honestly, the number of tests where we directly call internal methods instead of simulating HTML interactions (when possible) is kind of crazy. That’s not how bUnit is supposed to be used.

ScarletKuro commented:
Oh, so you already have that under the `// Test the UI` section, which does exactly what I'm talking about, it works perfectly.

The `// Test the method` section is absolutely unnecessary.  
Please don't do that in the future, it's worthless, as the method is not public. Therefore, always prefer HTML emulation over the programmatic approach.

I'm going to remove that part myself and merge it.

versile2 commented:
Definitely not necessary, but only way I could get the method to show
"tested"

On Sat, Apr 26, 2025, 1:29 PM Artyom M. ***@***.***> wrote:

> *ScarletKuro* left a comment (MudBlazor/MudBlazor#11083)
> <https://github.com/MudBlazor/MudBlazor/pull/11083#issuecomment-2832510840>
>
> Are you sure it's a good idea to test it by directly invoking
> GroupExpandClickAsync?
> Don’t we have an HTML counterpart with a handler like <MudIconButton ...
> ***@***.***" />, so we could just emulate clicking it
> instead? That’s always better.
> Honestly, the number of tests where we directly call internal methods
> instead of simulating HTML interactions (when possible) is kind of crazy.
> That’s not how bUnit is supposed to be used.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MudBlazor/MudBlazor/pull/11083#issuecomment-2832510840>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BDQDZ7HL6OSJHID2HWX2SR323PGCPAVCNFSM6AAAAABZVFBTRSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZSGUYTAOBUGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ScarletKuro commented:
> Definitely not necessary, but only way I could get the method to show "tested"

Where? I see in codecov it's invoked, the only line in that code that is not tested it this last line:
```
if (_expanded != GroupDefinition.Expanded)
   DataGrid._groupExpansionsDict.TryAdd(key, _expanded);
```

ScarletKuro commented:
wait why it got closed, I can't reopen it?

ScarletKuro commented:
I guess I will make new pr, ig I screwed something when I tried to modify this PR and now it doesn't let to do anything

ScarletKuro commented:
Well Only if you give me write access to `https://github.com/versile2/MudBlazor`

ScarletKuro commented:
Merging under https://github.com/MudBlazor/MudBlazor/pull/11243, sorry again for inconvenience.
